sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -133,37 +133,50 @@
   private final ValueProvider<Integer> numShardsProvider;
   private final boolean windowedWrites;
   private int maxNumWritersPerBundle;
+  // If set,this is used instead of DynamicDestinations.getSideInputs. This is used by the Beam
+  // translation code.
+  private final List<PCollectionView<?>> overriddenSideInputs;
+
+  private List<PCollectionView<?>> getSideInputs() {
+    return (overriddenSideInputs != null)
+        ? overriddenSideInputs
+        : getSink().getDynamicDestinations().getSideInputs();
```

Comment:
ok, I did modify the usage of side inputs in WriteFiles. Instead of expand() always querying the user's DynamicDestinations object for side inputs, WriteFiles captures the list of side inputs in its constructor. This makes the code cleaner IMO - the transform is responsible for its own side inputs, so this is no longer special code to make graph surgery work